GFAL2 and WebDAV testing

ThePhisch opened the pull request:
This PR covers issues #5962 and #6031 and implements tests for the GFAL2 and WebDAV RSE protocols. In addition to writing the test files, this means the following:

- Creating the RSE that rucio connects to
- initialising it within docker compose (in doing so, issue #5983 was solved)
- Integrating the tests into Github Actions (integration tests)

In addition, a docker image was created that runs a WebDAV server which rucio can connect to. It is this docker image that is referred to within the docker compose specifications. Of course, SSL certificates have been generated and baked into the image.

A final commit also handles #6041.

Commit:
Tests: GFAL2 Tests Written; Fix #5962

Wrote the tests for the GFAL2 RSE Protocol, basing them on the tests
written for the WebDAV RSE Protocol. The tests use pytest.

In these tests, GFAL2 makes use of the underlying WebDAV protocol. A
connection is set up to a WebDAV server. The tests also make use of curl
to communicate with the server. The tests do not clean up the testing
environment after running.

Tests were not inherited from WebDAV (and are not coupled in any other
way) because the interfaces provided by the WebDAV and GFAL2 protocol
implementations are different. For example, the function signatures are
different in many methods. In addition, the WebDAV implementation will
apply the path2pfn method by itself in many of its methods, whereas GFAL
requires inputs to habe already passed through a path2pfn method.

Commit:
Tests: SSL Certificates for WebDAV server; Fix #5962

Added lines to generate.sh to create the key pair for the WebDAV server
that the GFAL2 tests communicate with. GFAL2 uses the underlying WebDAV
protocol in the tests. In addition, GFAL2 requires the server it
communicates with to authenticate itself, hence the necessity for the
key pair.

Do not forget that you should not run the entire generate.sh file, only
portions of it (or else a new rucio CA will be generated).

The generated key pair has been baked into the WebDAV server image used
by the WebDAV and GFAL2 tests.

Also partially relevant for issue #6031 (testing the WebDAV RSE
Protocol).

Commit:
Tests: Add GFAL2, WebDAV Tests to Github Actions; Fix #5962 and #6031

Add the tests for RSE Protocols WebDAV and GFAL2 to the Github Action
integration tests.

Commit:
Tests: Decorative Change to make_temp_file

Added a type hint to clarify that this fixture teturns a Path object.

Commit:
Tests: Wrote WebDAV tests; Fix #6031

The WebDAV tests were completely rewritten to use pytest and make
maximum use of parametrisation and fixtures. Most methods of the WebDAV
protocol class are tested. Furthermore, timeout errors are specifically
tested for.

This implementation of tests does not make use of setup_ and teardown_class
methods to setup a testing folder before testing and delete the folder
after tests. This means that tests can be run concurrently, eg on github
with 3 runners. setup_method is called before each testcase and connects to
the WebDAV server with a fresh connection.

At the moment, there is no cleanup method after the tests.

Interactions with the server are also executed with calls to the command
line utility curl. This is done so that individual tests do not rely on
each other, and instead rely on curl to work.

Also added the WEB1 RSE (which is required for the tests) to the
activation script, including setting the distances to the other RSEs.

mlassnig commented on etc/docker/dev/docker-compose-storage-externalmetadata.yml:
```diff
@@ -1,4 +1,5 @@
-version: "3.5"
+version: '3.4'
```
going down a version number?

ThePhisch replied:
Fixed. I renamed the final commit (which concerns the `docker-compose`s) because it no longer fixes issue #5983, it only adds the `web1` container.

mlassnig commented on lib/rucio/tests/test_rse_protocol_gfal2_new.py:
```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
```
looks good, but maybe use a more descriptive name than `..._new` ? this suggests that the old file can go away?

ThePhisch replied:
yes, so i have been told. done.

mlassnig commented on tools/docker_activate_rses.sh:
```diff
@@ -89,6 +92,7 @@ rucio-admin account set-limits root XRD2 -1
 rucio-admin account set-limits root XRD4 -1
 rucio-admin account set-limits root SSH1 -1
+rucio-admin account set-limits root WEB1 -1
 
```
we should also connect the distances between them

Commit:
Tests: Add Web1 Containers; Fix #5962

The GFAL2 tests (and the WebDAV testing suite which they rely on)
requires a functioning WebDAV server. This server is run on a docker
container. This commit changes the relevant docker-composes so that the
"web1" container is spun up properly.

Commit:
Tests: Remove old GFAL2 and WebDAV Tests; Fix #5962

Removed the "_new" suffix from the newly written test files, thus
finally replacing the old test_rse_protocols with the new ones. Also
changed the names of the files called within integration_tests.yml.

test_rse_protocol_gfal2_new.py -> test_rse_protocol_gfal2.py
test_rse_protocol_webdav_new.py -> test_rse_protocol_webdav.py

Commit:
Tools: Split docker_activate_rses into activate and populate; Fix #6041

Split off the second half of `docker_activate_rses.sh` into
`docker_populate_rses.sh` which is sourced, provided the `--no-populate`
flag has not been set.

mlassnig reviewed (commented):
@ThePhisch @bari12 this is good to go. the only thing we need to decide is if we're ok with the try image being hosted by thephish's docker account, or if it needs to be made available on the rucio docker account?

ThePhisch commented:
> @ThePhisch @bari12 this is good to go. the only thing we need to decide is if we're ok with the try image being hosted by thephish's docker account, or if it needs to be made available on the rucio docker account?

I thought that the image will end up hosted on the rucio docker account

bari12 commented:
Yes, should definitely be the rucio docker account. So better add it to rucio/containers.

bari12 commented:
Push 😄
